### PR TITLE
[pdata] Change outcome of Metric data accessors misuse

### PR DESCRIPTION
**Pull request by dmitryax**
Change outcome of `pdata.Metric.<Gauge|Sum|Histogram|ExponentialHistogram>()` functions misuse. In case of type mismatch, they don't panic right away but return an invalid zero-initialized instance for consistency with other OneOf field accessors

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector/issues/4775

### CODE REVIEW

**Review by Aneurysm9 — APPROVED** [review]

File: model/internal/cmd/pdatagen/internal/base_fields.go [hunk 1]
```diff
@@ -64,10 +64,17 @@ const oneOfTypeAccessorHeaderTestTemplate = `func Test${structName}${originField
...
+// zero-initiazed instance of ${returnType}. Note that using such ${returnType} instance can cause panic.
+//
 // Calling this function on zero-initialized ${structName} will cause a panic.
 func (ms ${structName}) ${fieldName}() ${returnType} {
-	return new${returnType}((*ms.orig).${originOneOfFieldName}.(*${originStructType}).${originFieldName})
+	v := ms.orig.Get${originFieldName}()
+	if v == nil {
+		return ${returnType}{}
+	}
+	return new${returnType}(v)
```
Comment (bogdandrutu) [inline model/internal/cmd/pdatagen/internal/base_fields.go:77]: nit: This has an extra "if check" in the get. Try to do the type conversion here and use ok directly.

**Review by bogdandrutu — APPROVED** [review]

File: model/internal/cmd/pdatagen/internal/base_fields.go [hunk 2]
```diff
@@ -64,10 +64,17 @@ const oneOfTypeAccessorHeaderTestTemplate = `func Test${structName}${originField
 	assert.Equal(t, "", ${typeName}(1000).String())`
 
 const accessorsOneOfMessageTemplate = `// ${fieldName} returns the ${lowerFieldName} associated with this ${structName}.
-// Calling this function when ${originOneOfFieldName}Type() != ${typeName} will cause a panic.
+//
+// Calling this function when ${originOneOfFieldName}Type() != ${typeName} returns an invalid 
+// zero-initiazed instance of ${returnType}. Note that using such ${returnType} instance can cause panic.
```
Comment (mx-psi) [inline model/internal/cmd/pdatagen/internal/base_fields.go:69]:
```suggestion
// zero-initialized instance of ${returnType}. Note that using such ${returnType} instance can cause panic.
```

File: model/internal/cmd/pdatagen/internal/base_fields.go (hunk 1 above)
Comment (dmitryax) [reply model/internal/cmd/pdatagen/internal/base_fields.go]: done

File: model/internal/cmd/pdatagen/internal/base_fields.go (hunk 2 above)
Comment (dmitryax) [reply model/internal/cmd/pdatagen/internal/base_fields.go]: thanks for catching this. fixed

**Review by mx-psi — APPROVED** [review]

**Pull request merged**